packages/firestore/src/remote/remote_store.ts

```diff
@@ -480,29 +476,30 @@ export class RemoteStore implements TargetMetadataProvider {
         this.writePipeline.length > 0
           ? this.writePipeline[this.writePipeline.length - 1].batchId
           : BATCHID_UNKNOWN;
-      return this.localStore
-        .nextMutationBatch(lastBatchIdRetrieved)
-        .then(batch => {
-          if (batch === null) {
-            if (this.writePipeline.length === 0) {
-              this.writeStream.markIdle();
-            }
-          } else {
-            this.addToWritePipeline(batch);
-            return this.fillWritePipeline();
-          }
-        });
+      const batch = await this.localStore.nextMutationBatch(
+        lastBatchIdRetrieved
+      );
+
+      if (batch === null) {
+        if (this.writePipeline.length === 0) {
+          this.writeStream.markIdle();
+        }
+      } else {
+        this.addToWritePipeline(batch);
+        await this.fillWritePipeline();
+      }
+    }
+
+    if (this.shouldStartWriteStream()) {
+      this.startWriteStream();
     }
   }
 
   /**
-   * Returns true if we can add to the write pipeline (i.e. it is not full and
-   * the network is enabled).
+   * Returns true if we can add to the write pipeline (i.e. it is not full).
    */
   private canAddToWritePipeline(): boolean {
-    return (
-      this.isNetworkEnabled() && this.writePipeline.length < MAX_PENDING_WRITES
-    );
+    return this.writePipeline.length < MAX_PENDING_WRITES;
```

Comment:
`fillWritePipeline()` above is called by the syncEngine after a local write. Why would we want batches to be in the pipeline while the network is disabled?

Reply:
TBH it just makes more sense to me.  I don't see what we gain by throwing away the write pipeline when the network is disabled and then re-filling it (with the exact same writes) once it's enabled again.  We may as well just always be filling the write pipeline and then it will just consistently contain the next 10 writes to send to the write stream.

This is similar to how `listenTargets` persists across the network being disabled / re-enabled.

The only time we actually want to clear the write pipeline is on a user change, and so I reworked the handleUserChange code to do that explicitly.

Does this explanation sway you?  We may just have different conceptual models of what the write pipeline represents...  I'm willing to be convinced to reconsider.

Reply:
We need to make sure that the multi-tab code continues to throw away writes that were executed by other clients. This is simplified if we throw away the in-memory state and then re-fill the write pipeline.